### PR TITLE
feat(dynamic-sampling): Track metric every time a DSC is built

### LINKED ISSUE #3570: Investigate how many envelopes have no DSC

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds a new metric that collects the number of times the DSC is built by Relay when it was not found on a transaction that should contain the DSC.

The logic in this PR performs a semantic versioning comparison between the minimum version of an SDK that should support a DSC and the current SDK version of the SDK that sent the event.

Closes: https://github.com/getsentry/relay/issues/3570

#skip-changelog

### CODE REVIEW

File: relay-server/src/utils/dynamic_sampling.rs [hunk 1]
```diff
@@ -99,6 +215,75 @@ pub fn get_sampling_key(envelope: &Envelope) -> Option<ProjectKey> {
...
+
+    let Some(min_sdk_version) = SUPPORTED_SDK_VERSIONS.get(sdk_name.as_str()) else {
+        return;
+    };
+
+    match compare_versions(sdk_version, min_sdk_version) {
+        Ordering::Greater | Ordering::Equal => {
+            relay_statsd::metric!(
+                counter(RelayCounters::MissingDynamicSamplingContext) += 1,
+                sdk_name = sdk_name.as_str()
```
Comment (iambriccardo) [inline relay-server/src/utils/dynamic_sampling.rs]: This tag has technically unbounded cardinality but we know that the SDK names are not many.

File: relay-server/src/utils/dynamic_sampling.rs [hunk 2]
```diff
@@ -10,6 +12,120 @@ use relay_sampling::dsc::{DynamicSamplingContext, TraceUserContext};
...
 
 use crate::envelope::{Envelope, ItemType};
+use crate::statsd::RelayCounters;
+use once_cell::sync::Lazy;
+
+// List of minimum SDK versions that support Performance at Scale (aka Dynamic Sampling).
+// The list is defined here:
+// https://docs.sentry.io/product/performance/performance-at-scale/getting-started
+static SUPPORTED_SDK_VERSIONS: Lazy<HashMap<&'static str, &'static str>> = Lazy::new(|| {
+    let mut supported_sdk_versions = HashMap::new();
```
Comment (jjbayer) [inline relay-server/src/utils/dynamic_sampling.rs]: `HashMap::from([("key", "value"), ...])`

File: relay-server/src/utils/dynamic_sampling.rs [hunk 3]
```diff
@@ -99,6 +215,75 @@ pub fn get_sampling_key(envelope: &Envelope) -> Option<ProjectKey> {
     envelope.dsc().map(|dsc| dsc.public_key)
 }
 
+/// Compares two semantic versions.
+///
+/// This function is just temporary since it's used to compare SDK versions when an [`Event`] is
+/// received.
+fn compare_versions(version1: &str, version2: &str) -> std::cmp::Ordering {
```
Comment (jjbayer) [inline relay-server/src/utils/dynamic_sampling.rs]: Can we use the https://docs.rs/semver/latest/semver/ crate for this? It's already part of our dependency tree.

File: relay-server/src/utils/dynamic_sampling.rs (hunk 1 above)
Comment (jjbayer) [reply relay-server/src/utils/dynamic_sampling.rs]: You know that sdk_name is in SUPPORTED_SDK_VERSIONS, so it is bounded, right?

**Review by jjbayer — CHANGES_REQUESTED** [review]

File: relay-server/src/utils/dynamic_sampling.rs (hunk 1 above)
Comment (iambriccardo) [reply relay-server/src/utils/dynamic_sampling.rs]: Yeah in this case this is a good point, I forgot I was checking the presence in the map.

File: relay-server/src/utils/dynamic_sampling.rs (hunk 3 above)
Comment (iambriccardo) [reply relay-server/src/utils/dynamic_sampling.rs]: I imagined there was something, will use that, thanks!

File: relay-server/src/utils/dynamic_sampling.rs [hunk 4]
```diff
@@ -99,6 +213,42 @@ pub fn get_sampling_key(envelope: &Envelope) -> Option<ProjectKey> {
     envelope.dsc().map(|dsc| dsc.public_key)
 }
 
+/// Emits a metric when an [`Event`] is inside an [`Envelope`] without [`DynamicSamplingContext`].
+///
+/// This function is a temporary function which has been added mainly for debugging purposes. Our
+/// goal with this function is to validate how many times the DSC is not
```
Comment (Dav1dde) [inline relay-server/src/utils/dynamic_sampling.rs]: The end of the sentence is missing.

**Review by Dav1dde — COMMENTED** [review]
Can't we just emit a sentry error whenever we have a missing dsc at a certain sample rate, instead of doing this semver dance with a hardocded list of SDKs then just to have a metric which doesn't even include the SDK version?

**Comment by iambriccardo** [comment]
> Can't we just emit a sentry error whenever we have a missing dsc at a certain sample rate, instead of doing this semver dance with a hardocded list of SDKs?

We could, no strong opinions here. I didn't want to pollute our issues stream with this, since it might be very noisy. @jjbayer what do you think?

File: relay-server/src/utils/dynamic_sampling.rs [hunk 5]
```diff
@@ -99,6 +213,47 @@ pub fn get_sampling_key(envelope: &Envelope) -> Option<ProjectKey> {
...
+    if req.matches(&version) {
+        relay_statsd::metric!(
+            counter(RelayCounters::MissingDynamicSamplingContext) += 1,
+            sdk_name = sdk_name.as_str()
+        );
+
+        return true;
+    }
+
+    false
```
Comment (iambriccardo) [inline relay-server/src/utils/dynamic_sampling.rs]: Since I wanted easier testability of the function, I returned a `bool` here.

**Review by jjbayer — APPROVED** [review]
This is intended as a temporary metric right?

**Review by mitsuhiko — APPROVED** [review]

File: relay-server/src/utils/dynamic_sampling.rs [hunk 6]
```diff
@@ -326,4 +481,31 @@ mod tests {
...
+    #[test]
+    fn test_track_missing_dsc_with_supported_sdk_version() {
+        let mut event = mocked_event(EventType::Transaction, "bar", "2.0");
+        event.client_sdk = Annotated::new(ClientSdkInfo {
+            name: Annotated::new("sentry.python.flask".to_string()),
+            version: Annotated::new("1.7.3".to_string()),
+            ..Default::default()
+        });
+
+        let Some(sdk_name) = should_have_dsc(&event) else {
```
Comment (mitsuhiko) [inline relay-server/src/utils/dynamic_sampling.rs]: nit: just use unwrap here

File: relay-server/src/utils/dynamic_sampling.rs [hunk 7]
```diff
@@ -116,6 +251,26 @@ pub fn dsc_from_event(public_key: ProjectKey, event: &Event) -> Option<DynamicSa
...
+    let sdk_name = should_have_dsc(event);
+    match sdk_name {
+        Some(sdk_name) => {
+            relay_statsd::metric!(
+                counter(RelayCounters::MissingDynamicSamplingContext) += 1,
+                sdk_name = sdk_name.as_str(),
+                should_have_dsc = "true"
+            );
+        }
+        _ => {
```
Comment (mitsuhiko) [inline relay-server/src/utils/dynamic_sampling.rs]: nit: use `None`

**Pull request merged**